File: 0279-perfect-squares/0279-perfect-squares.go
/*
import (
    "math"
)

func numSquares(n int) int {
    table := make([]int, n+1)
    table[1] = 1
    
    for i := 1; i < n+1; i++ {
        minSquares := i
        for j := 1; j < int(math.Pow(float64(i), 0.5)) + 1; j++ {
            minSquares = min(minSquares, 1 + table[i-j*j])
        }
        table[i] = minSquares
    }
    
    return table[n]
}

*/
func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func numSquares(n int) int {
    cache := make(map[int]int, n)
    squares := make([]int, 0, n)
    
    for i := 1; i < n; i++ {
        cache[i*i] = 1
        squares = append(squares, i*i)
    }
    
    var numSquaresRecursive func(int) int
    numSquaresRecursive = func(num int) int {
        if count, ok := cache[num]; ok {
            return count
        }
        
        minSquares := num
        for _, s := range squares {
            if num < s {
                break
            }
            minSquares = min(minSquares, 1+numSquaresRecursive(num-s))
        }
        
        cache[num] = minSquares
        return minSquares
    }
    
    return numSquaresRecursive(n)
}

File: 0279-perfect-squares/0279-perfect-squares.py
"""
class Solution:
    def numSquares(self, n: int) -> int:
        num_squares = 0
        squares = []
        while 0 < n:
            tmp = n
            while n < tmp * tmp:
                tmp -= 1
            n -= tmp * tmp
            num_squares += 1
        return num_squares

class Solution:
    def numSquares(self, n: int) -> int:
        self.cache = {}
        squares = []
        for i in range(1, n):
            if i * i < n:
                squares.append(i * i)
                self.cache[i*i] = 1
            if i * i == n:
                return 1
        return self.num_squares_recursive(n, squares)
        
    def num_squares_recursive(self, n: int, squares: List[int]) -> int:
        if self.cache.get(n) != None:
            return self.cache[n]
        
        # all 1's, for example
        min_num = n
        for s in squares:
            if n < s:
                break
            if s == n:
                return 1
            min_num = min(min_num, 1 + self.num_squares_recursive(n-s, squares))
        
        self.cache[n] = min_num
        return min_num

class Solution:
    def numSquares(self, n: int) -> int:
        table = [0] * (n+1)
        table[1] = 1
        for i in range(1, n+1):
            min_squares = i
            for j in range(1, int(i**0.5)+1):
                min_squares = min(min_squares, 1 + table[i-j*j])
            table[i] = min_squares
        
        return table[n]

"""

class Solution:
    def numSquares(self, n: int) -> int:
        self.cache = {}
        self.squares = [i*i for i in range(1, int(n**0.5)+1)]
        return self.num_squares_recursive(n)
    
    def num_squares_recursive(self, n) -> int:
        if self.cache.get(n) != None:
            return self.cache[n]
        
        min_num = n
        for s in self.squares:
            if n < s:
                break
            min_num = min(min_num, 1+self.num_squares_recursive(n-s))
        
        self.cache[n] = min_num
        return min_num